apis/configwatch.go

```diff
@@ -181,3 +209,209 @@ func MonitorTLS(start time.Time, expiry time.Time, conf *config.Config) {
 		}
 	}
 }
+
+func (s *Server) getClientValidator(helloInfo *tls.ClientHelloInfo) func([][]byte, [][]*x509.Certificate) error {
+
+	log.Debug().Msgf("Inside get client validator - %v", helloInfo.Conn.RemoteAddr())
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		//added DNSName
+		opts := x509.VerifyOptions{
+			Roots:         s.CaCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		_, err := verifiedChains[0][0].Verify(opts)
+		if err != nil {
+			log.Error().Err(err).Msgf("certs verification failed")
+			return err
+		}
+
+		if len(s.DNSName) == 0 {
+			log.Info().Msgf("dnsname is undefined")
+			return nil
+		}
+		candidateName := toLowerCaseASCII(s.DNSName) // Save allocations inside the loop.
+		validCandidateName := validHostnameInput(candidateName)
+		for _, match := range verifiedChains[0][0].DNSNames {
+			// Ideally, we'd only match valid hostnames according to RFC 6125 like
+			// browsers (more or less) do, but in practice Go is used in a wider
+			// array of contexts and can't even assume DNS resolution. Instead,
+			// always allow perfect matches, and only apply wildcard and trailing
+			// dot processing to valid hostnames.
+			if validCandidateName && validHostnamePattern(match) {
+				if matchHostnames(match, candidateName) {
+					log.Debug().Msgf("Successfully matched matchHostnames")
+					return nil
+				}
+			} else {
+				if matchExactly(match, candidateName) {
+					log.Debug().Msgf("Successfully matched matchExactly")
+					return nil
+				} else if matchHostnamesWithRegexp(match, candidateName) {
+					log.Debug().Msgf("Successfully matched matchHostnamesWithRegexp")
+					return nil
+				}
+			}
+		}
```

Comment:
I think we should drop the complexity here and do the following.

1. Validate that each DNSName from the client cert is valid according to RFC
2. Match the DNSName from the client certificate against one or more regular expressions
3. If there is a match against the DNSName then return nil (match found / allow connection)
4. If there is not a match return err (match not found / connection not allowed)